### ISSUE TITLE
Dynamo hangs/takes long time for optimizers

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#1803]
Repro

1) Get this PR - https://github.com/pytorch/pytorch/pull/87492
2) Run this command - 
`python benchmarks/dynamo/huggingface.py --training --accuracy --device cuda --backend=eager --float32 --only=MobileBertForMaskedLM`

W/o optimizer, the above command takes few seconds to run.

**Comment by davidberard98** [comment]
on hf_T5_large (run these in the torchbench dir)

by default, these models run with the optimizer included.

```
python run.py hf_T5_large -t train -d cuda --torchdynamo inductor
```
^ this OOMs

```
python run.py hf_T5_large -t train -d cuda --torchdynamo inductor --torchinductor_cudagraph False
```
^ this takes ~30 minutes

... to remove the optimizer, go to `torchbenchmark/util/framework/huggingface/model_factor.py` and comment out `self.optimizer.step()`. Then re-run either of the above commands; this takes ~ a few minutes.

**Comment by mlazos** [comment]
So I investigated more, it looks like Dynamo tracing just takes a really long time when there are 1000s of parameters (one potential area for speedup I noticed is that we iterate over the whole graph every time we add a tensor variable). 

I have a more minimal repro here:

```python
import torch
import torch._dynamo as torchdynamo
import torch._inductor

model = torch.nn.Sequential(*[torch.nn.Linear(10, 10, True, device="cuda:0") for _ in range(1000)])
optimizer = torch.optim.Adam(model.parameters(), capturable=True)
x = torch.randn(10, 10, dtype=torch.float, device="cuda:0")
y = model(x)
torch.sum(y).backward()

@torchdynamo.optimize("eager")
def repro():
    optimizer.step()


repro()
```
After discussing with Yanbo, one potential solution I see is to avoid tracing the large loops over the parameter list in the optimizer by extracting them into functions and disabling dynamo on them. Then use the *for_each ops which will transfer the loop into the FX graph directly and avoid tracing each parameter indvidiually.

**Comment by mlazos** [comment]
Closing this, here are the final benchmark numbers from this [test script](https://gist.github.com/mlazos/7448766f62ccbca22da30efaff12a22f)

baseline commit (before optimizations): ee05f47bddfb97b4b292808543d928b3526fc0ca
optimization commit (includes all the above PRs): efdaddc09436d48b8f7a2792f5eb680714eaf42b

| param count | baseline trace time (s) | trace time with optimizations (s) |            
|-|-|-|                                                               
| 100 | 21.58 | 2.67  |                                                         
|200 | 90.312 | 5.42 |                                                      
|300 | 221.53  | 8.58 |                                                           
|400 | 429.84  | 12.27 |                                                              
|500 | 738.06   | 16.45|                                                       
|600 | 1191.47   | 20.57 |                                                          
|700 | 1755.94   | 26.13 |                                                        
|800 | 2447.60   | 31.00 |                                                       
|900 | 3344.54   | 37.01|                                                          
|1000 | 4464.85 | 42.74 |
